src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,49 +1,120 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+import java.util.Objects;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_ARRAY_SIZE = 10;
+    private Object[] array;
+    private int currentSize;
+
+    public ArrayList() {
+        currentSize = 0;
+        this.array = new Object[DEFAULT_ARRAY_SIZE];
+    }
+
+    private Object[] grow(Object[] oldArray) {
+        Object[] grownArray = new Object [oldArray.length + (oldArray.length >> 1)];
+        System.arraycopy(oldArray, 0, grownArray, 0, currentSize);
+        return grownArray;
+    }
 
     @Override
     public void add(T value) {
+        if (currentSize < array.length) {
+            array[currentSize] = value;
+            currentSize++;
+        } else {
+            array = grow(array);
+            this.add(value);
+        }
 
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index > currentSize || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index is out of list bounds");
+        }
+        if (index == currentSize) {
+            add(value);
+            return;
+        }
+        if (array.length <= currentSize - 1) {
+            array = grow(array);
+        }
+        System.arraycopy(array, index, array, index + 1, currentSize - index);
+        set(value, index);
+        currentSize++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        if (list == null || list.isEmpty()) {
+            return;
+        }
+        Object[] listArray = new Object[list.size()];
+        while (array.length <= currentSize + listArray.length) {
+            array = grow(array);
+        }
+        for (int i = 0; i < list.size(); i++) {
+            listArray[i] = list.get(i);
+        }
+        for (int i = 0; i < listArray.length; i++) {
+            this.add((T)listArray[i]);
+        }
```

Comment:
let's simplify this and use 1 loop and just call `add()` for each element in the list
size check is already done in `add()` method

Reply:
Yes, additional size check is useless.
But if simplify it to one loop, it will be impossible to add list to itself (arrayList.add(arrayList). It is not covered by tests.